pkg/hub/controllers/cluster/reconciler.go

```diff
@@ -458,3 +488,48 @@ func (r *Reconciler) InjectClient(c client.Client) error {
 	r.Client = c
 	return nil
 }
+
+func (r *Reconciler) handleClusterDeletion(cluster *hubv1alpha1.Cluster, ctx context.Context, req reconcile.Request) (bool, reconcile.Result, error) {
+	log := logger.FromContext(ctx)
+
+	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !controllerutil.ContainsFinalizer(cluster, clusterDeregisterFinalizer) {
+			controllerutil.AddFinalizer(cluster, clusterDeregisterFinalizer)
+			if err := r.Update(ctx, cluster); err != nil {
+				return true, reconcile.Result{}, err
+			}
+		}
+	} else {
+		// The object is being deleted
+		if controllerutil.ContainsFinalizer(cluster, clusterDeregisterFinalizer) && !cluster.Status.IsDeregisterInProgress && retryAttempts <= 3 {
```

Comment:
Cant we rely on the deregister status enum defined by hubv1alpha1 instead of the IsDeregisterInProgress flag?

Reply:
The `IsDeregisterInProgress` flag is set to `true` by controller to disable redundant event triggers and requeues of cluster reconciler in the controller.
The worker operator is simply utilizing the flag to retry 3-4 times before failing. Post which the controller will take care of removing the finalizer (after 10 mins)

We could use the enums, but the problem is to identify (at controller sied) if it is the first attempt or a retry which would not be possible if relying on the enums.